File: Python_projects/submission_002-mastermind-functions-master/submission_002-mastermind-functions-master/mastermind.py
import random

correct_digits_and_position = 0
correct_digits_only = 0
code = [0,0,0,0]

def num_gen():
    # this function generates the code and returns it """
    code = [0,0,0,0]
    #global code
    for i in range(4):
        value = random.randint(1, 8) # 8 possible digits
        while value in code:
            value = random.randint(1, 8)  # 8 possible digits
        code[i] = value
    print('4-digit Code has been set. Digits in range 1 to 8. You have 12 turns to break it.')
    return code
   
def user_input(code, answer):
    # this fuctions checks the input from the user to checks how many numbers in the guess are correct and returns whether the code is correct or not"""
    correct_digits_and_position = 0
    correct_digits_only = 0
    for i in range(len(answer)):
        if code[i] == int(answer[i]):
            correct_digits_and_position += 1
        elif int(answer[i]) in code:
            correct_digits_only += 1
    print('Number of correct digits in correct place:     '+str(correct_digits_and_position))
    print('Number of correct digits not in correct place: '+str(correct_digits_only))
    if correct_digits_and_position == 4:
        return True
    else:
        return False

def game_looper(code):
    # this function runs the main game loop, calls the other functions and inputs error checking"""
    correct = False
    turns = 0
    while not correct and turns < 12:
        answer = input("Input 4 digit code: ")
        if len(answer) > 4 or len(answer) < 4:
            print("Please enter exactly 4 digits.")
            continue
        correct = user_input(code, answer)
        turns += 1
        if correct == True:
            print('Congratulations! You are a codebreaker!')
        else:
            print('Turns left: '+str(12 - turns))

    print('The code was: '+str(code))

def run_game(): 
    # function runs game
    code = num_gen()
    game_looper(code) 

if __name__ == "__main__":
    run_game()

File: Java_projects/submission-102-java-003-mastermind-master/src/main/java/za/co/wethinkcode/mastermind/Mastermind.java
package za.co.wethinkcode.mastermind;
import java.util.*;

public class Mastermind {
    private final String code;
    private final Player player;
    private int match, intersect;
    private boolean isRunning;

    public Mastermind(CodeGenerator generator, Player player){
        this.code = generator.generateCode();
        this.player = player;
        this.match = 0;
        this.intersect = 1;
        this.isRunning = true;
        //System.out.println("fuck");
    }
    public Mastermind(){
        this(new CodeGenerator(), new Player());
    }

    public void runGame(){
        //TODO: implement the main run loop logic
        String user_guess;
        String prompt;

        System.out.println("4-digit Code has been set. Digits in range 1 to 8. You have 12 turns to break it.");
        while (this.isRunning) 
        {
            //get user guess
            //System.out.println("IN");

            user_guess = this.player.getGuess();
            //System.out.println(code);

            //System.out.println("OUT");
            // get two variables allowing for identification of user_guess
            this.countMatches(this.code, user_guess);
            this.countIntersect(this.code, user_guess);
            
            //update system (isRunning)
            System.out.printf("Number of correct digits in correct place: %d\n",
            this.match);
            System.out.printf("Number of correct digits not in correct place: %d\n",
            this.intersect);
            this.isRunning = (this.match == 4 || this.player.noGuesses())?false:true;

            //interface
            prompt = this.isRunning?"Turns left: "+this.player.getGuesses()
            :(this.match == 4)?"Congratulations! You are a codebreaker!":"No more turns left.";
            System.out.println(prompt);
            if (this.player.getGuesses() == 0)
            {
                System.out.println("The code was: "+ this.code);
            }
            else if(this.match == 4)
            {
                System.out.println("The code was: " + this.code);
            }

         //System.out.println(prompt);
         
        }
    }

    public void countMatches(String user_guess, String code)
    {
        int i = 0;
        

        for(int j = 0; j < 4; j++) {
            char str1 = user_guess.charAt(j);
            char str2 = code.charAt(j);  
            if ((str1 == str2)) {
                i++;
            }
        }
        
        this.match = i;
        
    }
    public void countIntersect(String user_guess, String code)
    {
        int count = 0;

        for(int i = 0; i < 4; i++)
        {
            for (int j = 0; j < 4; j++)
            {
                if(code.charAt(i) == user_guess.charAt(j) && i != j)
                {
                    count += 1;
                }
            }
        }
        this.intersect = count;
    }

    public static void main(String[] args){
        Mastermind game = new Mastermind();
        game.runGame();
    }
}
